Framework/PythonInterface/plugins/algorithms/PEARLTransfit.py

```diff
@@ -0,0 +1,371 @@
+        else:
+            noFiles = len(files) ** (-1)
+            CreateSingleValuedWorkspace(OutputWorkspace='scale', DataValue=noFiles)
+            Multiply(LHSWorkspace=firstFileName + '_3', RHSWorkspace='scale',
+                     OutputWorkspace=firstFileName + '_monitor')
+            DeleteWorkspace('scale')
+            DeleteWorkspace(firstFileName + '_3')
+
+    # ----------------------------------------------------------
+    # Define function for improved rebinning of the data
+    # ----------------------------------------------------------
+    def TransfitRebin(self, inputWS, outputWSName, foilType, divE):
+        ws2D = mtd[inputWS]
+        # Expand the limits for rebinning to prevent potential issues at the boundaries
+        startE = ResParamsDict[foilType + '_startE']
+        endE = ResParamsDict[foilType + '_endE']
+        startEp = 0.99 * startE
+        endEp = 1.01 * endE
+        CropWorkspace(InputWorkspace=ws2D, OutputWorkspace=ws2D, XMin=1000 * startEp, XMax=1000 * endEp)
+        numPts = np.int(((endEp - startEp) / divE) + 1)
+        xData_out = []
+        xData_in = ws2D.readX(0)
+        yData_in = ws2D.readY(0)
+        # Deals with issues in Mantid where Xdata mark start of bin, not true x position
+        # calculates the bin width and then takes middle of bin as x value
+        current_bin_widths = []
+        xActual = []
+        for x in range(0, len(xData_in) - 1):
+            current_bin_widths.append(xData_in[x + 1] - xData_in[x])
+            xActual.append(xData_in[x] + 0.5 * (xData_in[x + 1] - xData_in[x]))
+        # Make xData with uniform binning defined by divE
+        for j in range(numPts):
+            xData_out.append(1000 * (startEp + j * divE))
+        yData_out = [0] * (len(xData_out))
+        # Normalise output ydata accordingly based on surrounding values
+        yNorm = [0] * (len(xData_out))
+        for j in range(0, len(yData_in)):
+            currentBin = np.int((xActual[j] - startEp * 1000) / (divE * 1000))
+            scale1 = 1 - ((xActual[j] - xData_out[currentBin]) / (divE * 1000))
+            yData_out[currentBin] += yData_in[j] * scale1
+            yNorm[currentBin] += scale1
+            if currentBin < (len(xData_out) - 1):
+                yData_out[currentBin + 1] += yData_in[j] * (1 - scale1)
+                yNorm[currentBin + 1] += 1 - scale1
+        # Apply the normalisation, with a catch for any potential divide by zero errors
+        for i in range(len(yData_out)):
+            if yNorm[i] != 0:
+                yData_out[i] = yData_out[i] / yNorm[i]
+            else:
+                print('Empty bin')
+
+        outputWS = CreateWorkspace(DataX=xData_out, DataY=yData_out, NSpec=1, UnitX='meV')
+        CropWorkspace(InputWorkspace=outputWS, OutputWorkspace=outputWS, XMin=1000 * startE, XMax=1000 * endE)
+        RenameWorkspace(InputWorkspace=outputWS, OutputWorkspace=outputWSName)
+
+
+# Physical constants
+k = 1.38066e-23
```

Comment:
For some of these constants at least I think we should rely on the scipy.constants module. Just add this to top of the unit:

from scipy import constants

...and then can use constants.k, constants.e, constants.m_n, constants.h, constants.N_A instead of our own values. If you want to keep your own constant variable names and define k = constants.k then that's fine.

I think there are a few which are Pearl specific constants that you won't find in scipy eg flightpath\t0\t_stnd look like properties of the instrument

Please can you also move this section to the top of the class definition - just below L31

Reply:
One further comment on this. I just checked and the following constants aren't used in the file (maybe the original ResParams.py file was a general constants file Chris used across other algorithms):
m, mn, h, Na, flightpath, t0, t_stnd
...so think you can just delete these